viper/parser/parser.py

```diff
@@ -277,13 +277,31 @@ def __init__(self, vars=None, globals=None, sigs=None, forvars=None, return_type
         self.in_for_loop = set()
         # Count returns in function
         self.function_return_count = 0
+        # Current blockscope
+        self.blockscopes = []
 
     def set_in_for_loop(self, name_of_list):
         self.in_for_loop.add(name_of_list)
 
     def remove_in_for_loop(self, name_of_list):
         self.in_for_loop.remove(name_of_list)
 
+    def start_blockscope(self, blockscope_id):
+        if blockscope_id not in self.blockscopes:
+            self.blockscopes.append(blockscope_id)
+
+    def end_blockscope(self, blockscope_id):
+        if blockscope_id not in self.blockscopes:
+            return
+        # Remove all variables that have specific blockscope_id attached.
+        poplist = []
+        for name, var_record in self.vars.items():
+            if blockscope_id in var_record.blockscopes:
+                poplist.append(name)
+        for name in poplist: self.vars.pop(name)
+        # Remove blockscopes
+        self.blockscopes = self.blockscopes[self.blockscopes.index(blockscope_id):]
```

Comment:
Simplify to `self.blockscopes.pop(blockscope_id)`

Reply:
Pop would give the desired effect. If you have nested scopes, you want to slice awat all nested scopes as well.

Reply:
ah, that's an interesting thought. but the blockscopes should still both appear in order so I don't think this would happen e.g.

```python
for i in range(5): # "for" block started
    # "for" body
    if i == 3: # "if block started
         # "if" body
         return i
    # "if block ended
# "for" block ended
```

Reply:
Looking at it now pop would be 100%, was just my thought process at that moment... hahaha :smile_cat: